tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,288 @@
+            CONF_PASSWORD: PASSWORD,
+            CONF_ACCOUNT_NAME: ACCOUNT_NAME_FROM_USERNAME,
+        }
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "username_exists"
+
+    # Should fail, same USERNAME (flow)
+    result = await flow.async_step_user(
+        {CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["errors"] == {CONF_USERNAME: "username_exists"}
+
+
+async def test_abort_on_login_failed(hass: HomeAssistantType):
+    """Test when we have errors during login."""
+    flow = init_config_flow(hass)
+
+    with patch(
+        "pyicloud.base.PyiCloudService.authenticate",
+        side_effect=PyiCloudFailedLoginException(),
+    ):
+        result = await flow.async_step_user(
+            {CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD}
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["errors"] == {CONF_USERNAME: "login"}
+
+
+async def test_trusted_device(hass: HomeAssistantType, service: MagicMock):
+    """Test trusted_device step."""
+    flow = init_config_flow(hass)
+    flow.api = service
+
+    result = await flow.async_step_trusted_device()
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == CONF_TRUSTED_DEVICE
+
+
+async def test_trusted_device_success(hass: HomeAssistantType, service: MagicMock):
+    """Test trusted_device step success."""
+    flow = init_config_flow(hass)
+    flow.api = service
+
+    result = await flow.async_step_trusted_device({CONF_TRUSTED_DEVICE: 0})
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == CONF_VERIFICATION_CODE
+
+
+async def test_abort_on_send_verification_code_failed(
+    hass: HomeAssistantType, service_send_verification_code_failed: MagicMock
+):
+    """Test when we have errors during send_verification_code."""
+    flow = init_config_flow(hass)
+    flow.api = service_send_verification_code_failed
+
+    result = await flow.async_step_trusted_device({CONF_TRUSTED_DEVICE: 0})
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == CONF_TRUSTED_DEVICE
+    assert result["errors"] == {CONF_TRUSTED_DEVICE: "send_verification_code"}
+
+
+async def test_verification_code(hass: HomeAssistantType):
+    """Test verification_code step."""
+    flow = init_config_flow(hass)
+
+    result = await flow.async_step_verification_code()
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == CONF_VERIFICATION_CODE
+
+
+async def test_abort_on_validate_verification_code_failed(
+    hass: HomeAssistantType, service_validate_verification_code_failed: MagicMock
+):
+    """Test when we have errors during validate_verification_code."""
+    flow = init_config_flow(hass)
+    flow.api = service_validate_verification_code_failed
```

Comment:
In that case, I got 7 passing tests and 4 failing tests, with `AttributeError: 'NoneType' object has no attribute 'trusted_devices'` and `AttributeError: 'NoneType' object has no attribute 'validate_verification_code'` because `flow.api` hasn't been initialized ...

So to initialize it, I have to trigger `flow.async_step_user(self, user_input=None)` but then my test isn't really unitary !

Reply:
Are you ok with 7734677ede47e6a56a023a7516c27e8e2ae70823 ?

Reply:
We should test the config flow from start to finish and all the possible code paths between. So not starting with one of the entry points, ie the user step or the import step, wouldn't make sense.

Reply:
I’m fine with the explanation, in this case I understand, but otherwise tests should be unitary.